File: CRM-Server/app/schemas/reminder_rule.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

ReminderObjectType = Literal[
    "business_journey",
    "opportunity",
    "customer",
    "lead",
    "follow_up_task",
    "approval",
    "payment_plan",
]
ReminderTrigger = Literal["schedule", "change", "date"]


class ReminderCatalogChoice(BaseModel):
    value: str
    label: str


class ReminderCatalogField(BaseModel):
    value: str
    label: str
    operators: list[ReminderCatalogChoice]
    options: list[ReminderCatalogChoice] | None = None


class ReminderCatalogObject(BaseModel):
    object_type: ReminderObjectType
    label: str
    recipients: list[ReminderCatalogChoice]
    fields: list[ReminderCatalogField]


class ReminderRecipientUser(BaseModel):
    id: str
    name: str


class ReminderRuleWrite(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    object_type: ReminderObjectType
    trigger: ReminderTrigger
    version: Literal[1, 2] = 1
    status: str | None = None
    inactive_days: int | None = Field(None, ge=1, le=365)
    date_field: str | None = None
    offset_days: int | None = Field(0, ge=-30, le=30)
    require_no_new_activity: bool = False
    trigger_time: str | None = None
    conditions: list[dict[str, Any]] = Field(default_factory=list)
    recipients: list[str] = Field(..., min_length=1)
    message_title: str | None = Field(None, max_length=100)
    message_template: str = Field(..., min_length=1, max_length=1000)
    channels: list[Literal["in_app", "feishu"]] = Field(..., min_length=1)
    notify_once_per_window: bool = True


class ReminderRuleUpdate(ReminderRuleWrite):
    expected_revision: int = Field(..., ge=1)


class ReminderRuleEnabledUpdate(BaseModel):
    enabled: bool
    expected_revision: int = Field(..., ge=1)


class ReminderRuleView(ReminderRuleWrite):
    id: int
    revision: int
    enabled: bool
    sentence: str
    created_time: datetime
    last_modified_time: datetime


class ReminderRuleRunView(BaseModel):
    id: int
    rule_id: int
    object_type: str
    object_id: int
    message: str
    sent_count: int
    skipped_count: int
    created_time: datetime

File: CRM-Client/src/schemas/reminderRule.ts
import { z } from 'zod'

export const ReminderConditionSchema = z.object({
  field: z.string(),
  operator: z.string(),
  value: z.unknown().optional(),
})
export const ReminderCatalogChoiceSchema = z.object({ value: z.string(), label: z.string() })
export const ReminderCatalogFieldSchema = ReminderCatalogChoiceSchema.extend({
  operators: z.array(ReminderCatalogChoiceSchema),
  options: z.array(ReminderCatalogChoiceSchema).nullish(),
})
export const ReminderObjectCatalogSchema = z.object({
  object_type: z.enum(['business_journey', 'opportunity', 'customer', 'lead', 'follow_up_task', 'approval', 'payment_plan']),
  label: z.string(),
  recipients: z.array(ReminderCatalogChoiceSchema),
  fields: z.array(ReminderCatalogFieldSchema),
})
export const ReminderRecipientSchema = z.object({ id: z.string(), name: z.string() })


export const ReminderRuleSchema = z.object({
  id: z.number(),
  name: z.string(),
  object_type: z.enum(['business_journey', 'opportunity', 'customer', 'lead', 'follow_up_task', 'approval', 'payment_plan']),
  version: z.number().default(1),
  trigger: z.enum(['schedule', 'change', 'date']),
  status: z.string().nullable().optional(),
  inactive_days: z.number().nullable().optional(),
  date_field: z.string().nullable().optional(),
  offset_days: z.number().nullable().optional(),
  require_no_new_activity: z.boolean().default(false),
  trigger_time: z.string().nullable().optional(),
  conditions: z.array(ReminderConditionSchema).default([]),
  recipients: z.array(z.string()),
  message_title: z.string().nullable().optional(),
  message_template: z.string(),
  channels: z.array(z.string()),
  notify_once_per_window: z.boolean().default(true),
  enabled: z.boolean(),
  revision: z.number().int().positive(),
  sentence: z.string(),
  created_time: z.string(),
  last_modified_time: z.string(),
})

export const ReminderRuleRunSchema = z.object({
  id: z.number(),
  rule_id: z.number(),
  object_type: z.string(),
  object_id: z.number(),
  message: z.string(),
  sent_count: z.number(),
  skipped_count: z.number(),
  created_time: z.string(),
})

export type ReminderCondition = z.infer<typeof ReminderConditionSchema>
export type ReminderObjectCatalog = z.infer<typeof ReminderObjectCatalogSchema>
export type ReminderRule = z.infer<typeof ReminderRuleSchema>
export type ReminderRuleRun = z.infer<typeof ReminderRuleRunSchema>
export type ReminderRecipient = z.infer<typeof ReminderRecipientSchema>
